### DISCUSSION TITLE
Is it possible to cancel/abort a solve in progress?

### DISCUSSION THREAD

**Discussion by qwertie** [ERGO-Code/HiGHS#1403] [Q&A] [answered]
lpsolve, for example, has `put_abortfunc` to poll whether a solve should continue. What's the HiGHS equivalent?

**Comment by jajhall** [comment, marked as answer]
It's not currently possible to abort HiGHS in this. However, we have a callback in a development branch that will permit this.

**Reply by qwertie** [reply]
Is there a way to fake it, e.g. can it be asked to do one iteration, then one more, then one more, until it finishes or we want to cancel?

**Reply by jajhall** [reply]
Technically yes, but there is a meaningful overhead of restarting the simplex solver from the current basis. There's no refactorization, but some extra solves. If you run HiGHS iteration by iteration like this, it might be as slow as lpsolve 😉

I'm puzzled as to why you want to do this. If it's because you're used to having to abort lpsolve, you may find that it's unnecessary with HiGHS as it's so much faster.

**Reply by qwertie** [reply]
We currently have a solve that takes 20 minutes on HiGHS (an improvement over the 10 hours used by Apache Math) and uses 19 GB of RAM. Starting a second solve without stopping the first one is right at the edge of what the server can do (I have a plan for simplifying the constraints, but it'll probably take a few weeks and we want something sooner). So, how does one go about limiting the number of iterations and then restarting from the current basis?

**Reply by jajhall** [reply]
Are you solving a linear programming (LP) problem, or mixed integer programming problem?

If you're solving an LP, are you using interior point or simplex? If you don't know, then it's simplex by default.

**Reply by qwertie** [reply]
It's a pure LP running in the default mode (so simplex - dual simplex?).

**Reply by jajhall** [reply]
So, set the option `simplex_iteration_limit`

**Reply by qwertie** [reply]
So, something like this (or by setting `time_limit` instead)? Is it necessary to increment the limit each time?

~~~cs
    HighsSolution? solution = null;

    int maxIterations = 1;
    solver.setIntOptionValue("simplex_iteration_limit", maxIterations);
                    
    for (;;) {
        status = solver.run(); // Highs_run

        var modelStatus = solver.GetModelStatus();
        if (cancelRequested || modelStatus is not (HighsModelStatus.kSolutionLimit or 
            HighsModelStatus.kIterationLimit or HighsModelStatus.kTimeLimit))
            break;

        ++maxIterations;
        solver.setIntOptionValue("simplex_iteration_limit", maxIterations);
    }

    if (status != HighsStatus.kError)
        solution = solver.getSolution();
~~~

**Reply by jajhall** [reply]
Sorry, I've just tried it out, and what I suggested doesn't work. 

TBH It's such a bizarre way of running an LP solver, that I'd not tried it before.

Unfortunately, if HiGHS terminates due to the iteration (or time limit) being reached, it just bails out, and doesn't save the data required for a re-start. Hence it would just start from scratch again

It could be fixed, but it's not quick to do

Better to wait for our callbacks

Since you're calling from C++, if you're building HiGHS from source, you could try. If so, let me know and I'll explain how the callbacks work. They are so new that they aren't documented

**Reply by qwertie** [reply]
I'm calling from C# and using the Windows/Linux [prebuilt files](https://github.com/JuliaBinaryWrappers/HiGHS_jll.jl/releases) like lib\libhighs.so.1.5.3 in `HiGHS.v1.5.3.x86_64-linux-gnu-cxx11.tar.gz`, though I have VC++, gcc (WSL) and cmake so maybe I could build from source. C# evidently supports C callbacks well enough, e.g. the one for lpsolve is defined like

        [DllImport("lpsolve55", ...)]
        public static extern void put_abortfunc(IntPtr lp, ctrlcfunc newctrlc, int ctrlchandle);

        public delegate byte ctrlcfunc(IntPtr lp, int userhandle);

corresponding to the C function

~~~cpp
typedef int    (__WINAPI lphandle_intfunc)(lprec *lp, void *userhandle);
...
__EXTERN_C {
...
void __EXPORT_TYPE __WINAPI put_abortfunc(lprec *lp, lphandle_intfunc newctrlc, void *ctrlchandle);
...
}
~~~

(Hmm, passing `int` to `void*` shouldn't actually work on x64. Maybe the calling convention stops it from crashing)

**Reply by jajhall** [reply]
Ah, extending out callbacks to our Python API and C (and hence, C#) API is something that we know we'll need to do, but it won't happen soon, sorry.
